### ISSUE TITLE
Intercept ICY onMetadataChanged calls

### ISSUE THREAD

**Issue by 6rism0** [androidx/media#407]
This question is based on an idea from a closed issue thread (see #258)

---

In order to update the media item (audio stream) metadata from a different source other than the initial or in-stream metadata, we've used a `ForwardingPlayer` like suggested, see https://github.com/androidx/media/issues/258#issuecomment-1453449978

So far it works quite well, but somehow the ICY metadata keeps getting delivered to listers that attach to the MediaController build from the Session which uses the ForwardingPlayer. (The notification e.g. only shows the metadata I propagate and not the ICY metadata)

I assumed that the ForwardingPlayer subclass which is passed to new MediaSession.Builder(...), would be the MediaController I get with MediaController.Builder(context, mediaSessionToken), but that doesn't seem to be the case.

--- 

The icy metadata is quite limited and we try to present only the richer information from an external API, but we might need the icy metadata as a trigger, though. So we can't totally block any icy onMediaMetadataChanged calls from within the actual player.

The idea based on the suggestion was, catch onMediaMetadataChanged calls in the ForwardingPlayer (maybe use them as a trigger to fetch 'richer' metadata from an api) and then call onMediaMetadataChanged with custom metadata on every attached listener.

---

#### ForwardingPlayer
```
class MetadataForwardingPlayer(wrapped: Player) : ForwardingPlayer(wrapped) {

    override val fromPlayerListener: Player.Listener = object : Player.Listener {

          override fun onMediaMetadataChanged(mediaMetadata: MediaMetadata) {
              val customMetadata: MediaMetadata = getMetadataFromApi()
              updateMetadata(customMetadata)
          }
  
          override fun onMetadata(metadata: Metadata) = Unit

    }

    private val listenerSet: MutableSet<Player.Listener> = mutableSetOf()

    init {
        wrapped.addListener(fromPlayerListener)
    }

    override fun getMediaMetadata(): MediaMetadata {
        return MetadataStorage.latestMediaMetadata
    }

    override fun addListener(listener: Player.Listener) {
        super.addListener(listener.also { listenerSet.add(it) })
    }

    override fun removeListener(listener: Player.Listener) {
        super.removeListener(listener.also { listenerSet.remove(it) })
    }

    private fun updateMetadata(mediaMetadata: MediaMetadata) {
        MetadataStorage.latestMediaMetadata = mediaMetadata
        listenerSet.forEach { it.onMediaMetadataChanged(mediaMetadata) }
    }
    
    (...)
}

```

#### Session 
```
class SessionService : MediaLibraryService() {

    override fun onCreate() {
        mediaSession = MediaLibrarySession.Builder(this, MetadataForwardingPlayer(actualPlayer), callback).build()
        (...)
    }
    (...)
}
```

#### Controller
```
private val mediaSessionToken: SessionToken = SessionToken(context, ComponentName(context, sessionServiceClass))

private val mediaControllerFuture = MediaController.Builder(context, mediaSessionToken).buildAsync()
```

**Comment by icbaker** [comment]
This is the problematic line:

```
override fun addListener(listener: Player.Listener) {
    super.addListener(listener.also { listenerSet.add(it) })
}
```

Because if I have a handle to your `MetadataForwardingPlayer` and I call `addListener` then the listener I register is registered **directly on the delegate `Player`**, so any calls that the underlying player makes to `onMediaMetadataChanged` will propagate out to my listener.

Instead of directly passing this listener to the delegate, you will need to wrap it in a `Player.Listener` implementation that drops all `onMediaMetadataChanged` calls from the delegate player on the floor. If you do this, you need to be careful of how you handle `removeListener` too, in terms of forwarding it to the delegate (because it's expected to pass the same `Listener` instance to both `addListener` and `removeListener`). See how `ForwardingPlayer.ForwardingListener` overrides `equals` in order to make this work:
https://github.com/androidx/media/blob/e268a2e1e2ffac4a13bb655e8f924b65466a20be/libraries/common/src/main/java/androidx/media3/common/ForwardingPlayer.java#L1124-L1135

**Comment by 6rism0** [comment]
Yes right, now I get it.
Really appreciate the help, thank you!

**Comment by 6rism0** [comment]
Reopening this issue, because I've stumbled upon some strange behaviour regarding bluetooth devices and metadata updates. This might be worth a separate ticket, but as its closely related to the things described above so I'll start with reopening this one.

---

The implementation described above works perfectly fine for my use case, the only issue I've got is related to bluetooth devices that are able to display metadata (e.g. car consoles). It seems like those devices do not receive the custom metadata calls we trigger, they get stuck after the first metadata update they receive and only update when the player state changes.

Is there a way to trigger custom metadata updates on bluetooth devices as well?
